### ISSUE TITLE
Review modelling of generic/template types

### ISSUE THREAD

**Issue by NicolasAnquetil** [moosetechnology/Famix#508]
Currently there is:
- TWithParameterizedTypes which seems to represent generic (or template) class declaration (eg: Collection<T>)
- TParameterizedType which seems to represent "instanciation" of a generic/template with concrete type arguments (eg Collection<String>)
- TParameterType which I am not sure what it is (no relationship with any of the other ...)
- TWithParameterizedTypeUsers which I am not sure what it is
- TParameterizedTypeUsers which I am not sure what it is

We need 4 things:
1. a type can be a generic/template (Collection<T>)
2. a type can be a parameter in a generic/template ("T" in the above)
3. a type can be an "instanciation" of a generic/template (Collection<String>)
4. a type can be an argument in the instanciation of a generic/template ("String" in the above)

I propose that:
1. a generic/template is called a TGenericType, TTemplateType, or TWithParameterTypes
2. a parameter type is called TGenericParameterType, TTemplateParameterType, or TParameterType
3. an instanciation of a generic/template is a TGenericInstanciationType, TTemplateInstanciationType, or TParameterizedType (same name but different semantics than what we currently have)
4. an argument in an instanciation of a generic/template is just a Type

![image](https://user-images.githubusercontent.com/14889146/211142242-cc040297-d2ab-4317-8e28-a227df46d96e.png)

We may or may not represent the relation between a TTemplateParameterType and the concrete Types used as argument in a TTemplateInstanciationType

**Comment by NicolasAnquetil** [comment]
Note in java a ParameterizableClass has ParameterTypes (has suggested above), but instead of creating it in the generator, it seems to be done manually
```Smalltalk
FamixJavaParameterizableClass >> parameters
	<FMProperty: #parameters type: #FamixJavaParameterType> <multivalued> <derived>
	<FMComment: 'Parameter types of this class.'>
	
	^self types select: [:each | each isParameterType]
```
and the opposite link does not exist in FamixJavaParameterType (so the "derived" relation above is useless)

**Comment by NicolasAnquetil** [comment]
note: related to #505

**Comment by NicolasAnquetil** [comment]
A bit more complex than that, because some other entities are parameterizable, notably methods (Java and C++) or functions (C/C++).
So we would need an abstract ParamerizedEntity

The solution could be something like this:
![image](https://user-images.githubusercontent.com/14889146/211770467-805d9fec-a000-448d-968b-b0b34795b289.png)

**Comment by badetitou** [comment]
Hello, 

The issue https://github.com/moosetechnology/Famix/issues/511 killed the `TWhitMethods` of `FamixJavaParameterizedType`. Which in turn killed VerveineJ.

IMHO we should really think about the process to not deliver Moose image that does not work with the last VerveineJ.
To do so we can use the CI to perform continuous delivery of FamixJava sync with the last VerveineJ. If you agree, I can help set up such a thing. 

About this issue (#508). I think we have to think now about the other traits `TWithMethods`, `TWithAttributes` and so on. Our metamodel is not enough ready.

**Comment by ClotildeToullec** [comment]
Status of the discussion after a meeting: 
![Generic drawio(1)](https://user-images.githubusercontent.com/39184695/221546880-e906e722-2012-4fd3-b94a-479462c3b205.png)

Remaining points to discuss: 
- The `GenericEntity`  of a concretisation could be obtained via its `parameterType` but this way the concretisation would not have a target.
- An entity can be generic and concrete at the same time, like in: `MyClass <T,Boolean>  myVar`
- A `TGenericEntity`has 1..n `ParameterType`. But this trait is used by `Type`, `Method`and `Function`so there can be 0 `parameterType` if the entity is not generic

**Comment by ClotildeToullec** [comment]
When several concretisations happen: a `TGenericEntity`has 0..* `TConcretisations`. One for each entity concretising one of its `TParameterType`.
In this code snippet:
```java
public class MyClass<T,U> {};

// A & B
MyClass<A,B> myVar1 = new MyClass<A,B>();
// C & D
MyClass<C,D> myVar2 = new MyClass<C,D>();
// A & D
MyClass<A,D> myVar3 = new MyClass<A,D>();
```

- On each line, we have 2 concretisations with the same source and target but different `parameterType` (T & U) and `concreteType` (A & B for `var1`). 
- The 2 concretisations of each line are "simultaneous" but we have no direct way of knowing it. 
- For `var1` and `var2`, we have no way of knowing that the simultaneous concretisations are A & B then C & D and not A & D and C & B.
- For `var1`and `var3`, there are 4 concretisations. 2 of them have different source but same target, same parameterType (T) and same concreteType (A)

**Comment by badetitou** [comment]
I discover that this PR will solve a problem with exception we have today:

```java
ArrayList<Exception>
```

can not be represented in our model because `Exception` is not `TParameterizedTypeUser`. But since it is a `TType`, proposal will fix this issue (I hope so)

**Comment by anneetien** [comment]
Fix with #508
